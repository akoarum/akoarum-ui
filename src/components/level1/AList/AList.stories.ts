import { storiesOf } from '@storybook/vue'
import { withKnobs, boolean } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import AList from './AList.vue'

const items = [
  { id: 0, text: 'アメリカンショートヘア' },
  { id: 1, text: 'マンチカン' },
  { id: 2, text: 'ノルウェイジャンフォレストキャット' },
  { id: 3, text: 'スコティッシュフォールド' }
]

storiesOf('level1/AList', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Unordered', () => ({
    components: { AList },
    props: {
      ordered: { default: boolean('ol', false) }
    },
    data() {
      return {
        items
      }
    },
    template: `
      <AList :items="items" :ordered="ordered" />
    `
  }), { info: {} })
  .add('Ordered', () => ({
    components: { AList },
    props: {
      ordered: { default: boolean('ol', false) }
    },
    data() {
      return {
        items
      }
    },
    template: `
      <AList :items="items" :ordered="ordered" visual-ordered />
    `
  }), { info: {} })
