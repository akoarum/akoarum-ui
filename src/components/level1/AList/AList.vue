<script lang="ts">
import { CreateElement, VNode, PropType } from 'vue'
import { defineComponent, computed } from '@vue/composition-api'
import { ListItem } from '~/types/ListItem'

interface Props {
  items: ListItem[]
  ordered: boolean
  visualOrdered: boolean
}

export default defineComponent({
  props: {
    items: { type: Array as PropType<ListItem[]>, required: true },
    ordered: { type: Boolean },
    visualOrdered: { type: Boolean }
  },
  setup(props: Props) {
    const styledClasses = computed<string[]>(() => {
      const classes: string[] = ['aList']
      if (props.visualOrdered) classes.push('-ordered')
      return classes
    })

    return { styledClasses }
  },
  render(createElement: CreateElement): VNode {
    const Items = this.items.map(item =>
      createElement(
        'li',
        {
          class: 'aList__item',
          props: {
            key: item.id
          }
        },
        item.text
      )
    )
    return createElement(
      this.ordered ? 'ol' : 'ul',
      {
        class: this.styledClasses
      },
      Items
    )
  }
})
</script>

<style lang="scss" scoped>
.aList {
  counter-reset: item;
  font-size: 14px;
  line-height: 1.6;

  &.-ordered {
    .aList__item {
      padding-left: 18px;

      &::before {
        left: 0;
        top: 0;
        content: counter(item) '.';
        border-radius: 0;
        width: auto;
        height: auto;
        background: transparent;
      }
    }
  }
}

.aList__item {
  counter-increment: item;
  position: relative;
  padding-left: 16px;

  &::before {
    position: absolute;
    left: 2px;
    top: 0.6em;
    content: '';
    border-radius: 50%;
    width: 5px;
    height: 5px;
    background: $color-primary;
  }
}

.aList__item + .aList__item {
  margin-top: 4px;
}
</style>
