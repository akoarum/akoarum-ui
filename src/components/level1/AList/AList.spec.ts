import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import AList from './AList.vue'

describe('AList', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(AList, {
      propsData: {
        items: [
          { id: 0, text: 'アメリカンショートヘア' },
          { id: 1, text: 'マンチカン' },
          { id: 2, text: 'ノルウェイジャンフォレストキャット' },
          { id: 3, text: 'スコティッシュフォールド' }
        ]
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it('styledClasses: 見た目はpropで制御できる', () => {
      wrapper.setProps({ visualOrdered: true })
      expect(wrapper.vm.styledClasses).toEqual(['aList', '-ordered'])
    })
  })
})
