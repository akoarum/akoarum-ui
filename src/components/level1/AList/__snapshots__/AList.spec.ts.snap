// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AList initialized snapshot 1`] = `
<ul
  class="aList"
>
  <li
    class="aList__item"
  >
    アメリカンショートヘア
  </li>
  <li
    class="aList__item"
  >
    マンチカン
  </li>
  <li
    class="aList__item"
  >
    ノルウェイジャンフォレストキャット
  </li>
  <li
    class="aList__item"
  >
    スコティッシュフォールド
  </li>
</ul>
`;
